core/query/mixingQuery.ml

```diff
@@ -0,0 +1,1551 @@
+    | Q.Project (r, label) ->
+        let rec project (r, label) =
+          match r with
+            | Q.Record fields ->
+              assert (StringMap.mem label fields);
+              StringMap.find label fields
+            | Q.If (c, t, e) ->
+              Q.If (c, project (t, label), project (e, label))
+            | Q.Var (_x, Types.Record row) ->
+              let field_types =  Q.field_types_of_row row in
+              assert (StringMap.mem label field_types);
+              Q.Project (r, label)
+            | _ -> query_error ("Error projecting label %s from record: %s") label (string_of_t r)
+        in
+        retn in_dedup (project (norm false env r, label))
+    | Q.Erase (r, labels) ->
+        let rec erase (r, labels) =
+          match r with
+          | Q.Record fields ->
+            assert (StringSet.for_all
+                      (fun label -> StringMap.mem label fields) labels);
+            Q.Record
+              (StringMap.fold
+                 (fun label v fields ->
+                   if StringSet.mem label labels then
+                     fields
+                   else
+                     StringMap.add label v fields)
+                 fields
+                 StringMap.empty)
+          | Q.If (c, t, e) ->
+            Q.If (c, erase (t, labels), erase (e, labels))
+          | Q.Var (_x, Types.Record row) ->
+            let field_types = Q.field_types_of_row row in
+            assert (StringSet.subset labels (Q.labels_of_field_types field_types));
+            Q.Erase (r, labels)
+          | _ -> query_error "Error erasing from record"
+        in
+        erase (norm false env r, labels)
+    | Q.Variant (label, v) -> Q.Variant (label, norm false env v)
+    | Q.Apply (f, xs) as _orig ->
+      apply in_dedup env (norm false env f, List.map (norm false env) xs)
+    | Q.For (_, gs, os, u) as _orig ->
+        (* Debug.print ("norm.For: " ^ Q.show _orig); *)
+        let rec reduce_gs env os_f body = function
+        | [] ->
+          begin
+            match norm in_dedup env body with
+            | Q.For (_, gs', os', u') ->
+                reduce_gs env (os_f -<- (fun os'' -> os'@os'')) u' gs'
+            (* this special case allows us to hoist a non-standard For body into a generator *)
+            | Q.Prom _ as u' ->
+                let z = Var.fresh_raw_var () in
+                let tyz =
+                  Q.type_of_expression u'
+                  |> TypeUtils.element_type
+                in
+                let vz = Q.Var (z, tyz) in
+                (* Debug.print ("norm.For fresh var: " ^ Q.show vz); *)
+                Q.reduce_for_source env (z, u', tyz) (fun env' os_f' ->
+                  Q.For (None, [], List.map (norm false env') (os_f' (os_f [])),
+                    norm in_dedup env' (Q.Singleton vz)))
+            | u' ->
+              (* Q.For (None, [], List.map (norm false env) (os_f []), u') *)
```

Comment:
Is this debugging code still needed?  It looks like the initial commented-out version is equivalent.

Reply:
Indeed, much of that debugging code isn't needed.